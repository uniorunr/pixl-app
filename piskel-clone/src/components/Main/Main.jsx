import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Tools from './Tools/Tools';
import Frames from './Frames/Frames';
import Canvas from './Canvas/Canvas';
import Preview from './Preview/Preview';
import ColorSelect from './ColorSelect/ColorSelect';
import Layers from './Layers/Layers';
import './Main.scss';

const getWindowSize = () => {
  const windowWidth = ((window.innerWidth / 128) * 0.5).toFixed(0) * 128;
  const windowHeight = ((window.innerHeight / 128) * 0.8).toFixed(0) * 128;
  return windowWidth < windowHeight ? windowWidth : windowHeight;
};

class Main extends Component {
  constructor() {
    super();

    this.state = {
      width: getWindowSize(),
      height: getWindowSize(),
      pixelsPerCanvas: 64,
      frames: [],
      framesData: JSON.parse(sessionStorage.getItem('framesData')) || {},
      layerKeys: [],
      activeLayer: +sessionStorage.getItem('activeLayer') || 0,
      primaryColor: sessionStorage.getItem('primaryColor') || '#000000',
      secondaryColor: sessionStorage.getItem('secondaryColor') || '#ffffff',
    };
  }

  componentDidMount = () => {
    window.addEventListener('resize', this.updateDimensions);
  };

  updateFrames = (frames) => {
    this.setState({
      frames: [...frames],
    });
  };

  updateLayerKeys = (keys) => {
    this.setState({
      layerKeys: [...keys],
    });
  };

  updateActiveLayer = (index) => {
    this.setState({
      activeLayer: index,
    });
  };

  updateDimensions = () => {
    this.setState({
      width: getWindowSize(),
      height: getWindowSize(),
    });
  };

  handlePixelsPerCanvas = (event) => {
    this.setState({
      pixelsPerCanvas: +event.currentTarget.value,
    });
  };

  updateColor = (color, isPrimary) => {
    if (isPrimary) {
      this.setState({
        primaryColor: color,
      });
    } else {
      this.setState({
        secondaryColor: color,
      });
    }
  };

  render() {
    const {
      width,
      height,
      pixelsPerCanvas,
      frames,
      primaryColor,
      secondaryColor,
      framesData,
      layerKeys,
      activeLayer,
    } = this.state;

    const {
      toolsData,
      currToolId,
      updateCurrentTool,
      framesShortcuts,
    } = this.props;

    return (
      <main className="main">
        <section>
          <Tools
            toolsData={toolsData}
            updateCurrentTool={updateCurrentTool}
            currToolId={currToolId}
          />
          <ColorSelect
            primaryColor={primaryColor}
            secondaryColor={secondaryColor}
            updateColor={this.updateColor}
          />
        </section>
        <Frames
          framesArray={frames}
          framesData={framesData}
          layerKeys={layerKeys}
          activeLayer={activeLayer}
          updateFrames={this.updateFrames}
          framesShortcuts={framesShortcuts}
        />
        <Canvas
          width={width}
          height={height}
          pixelsPerCanvas={pixelsPerCanvas}
          currToolId={currToolId}
          primaryColor={primaryColor}
          framesArray={frames}
          framesData={framesData}
          layerKeys={layerKeys}
          activeLayer={activeLayer}
          secondaryColor={secondaryColor}
          updateColor={this.updateColor}
        />
        <section className="right-section">
          <Preview
            framesArray={frames}
            handlePixelsPerCanvas={this.handlePixelsPerCanvas}
          />
          <Layers
            framesArray={frames}
            framesData={framesData}
            updateLayerKeys={this.updateLayerKeys}
            updateActiveLayer={this.updateActiveLayer}
          />
        </section>
      </main>
    );
  }
}

Main.propTypes = {
  toolsData: PropTypes.instanceOf(Object).isRequired,
  currToolId: PropTypes.string.isRequired,
  updateCurrentTool: PropTypes.func.isRequired,
  framesShortcuts: PropTypes.instanceOf(Object).isRequired,
};

export default Main;
