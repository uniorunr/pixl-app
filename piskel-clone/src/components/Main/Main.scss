.main {
  display: flex;
  justify-content: space-between;
  align-items: center;

  .canvas-section__canvas {
    border: 2px solid black;
  }
}
