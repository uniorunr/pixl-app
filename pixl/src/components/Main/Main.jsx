import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../../actions/actions';
import Tools from './Tools/Tools';
import Frames from './Frames/Frames';
import Canvas from './Canvas/Canvas';
import Preview from './Preview/Preview';
import ColorSelect from './ColorSelect/ColorSelect';
import Layers from './Layers/Layers';
import './Main.scss';

const getWindowSize = () => {
  const windowWidth = ((window.innerWidth / 128) * 0.5).toFixed(0) * 128;
  const windowHeight = ((window.innerHeight / 128) * 0.8).toFixed(0) * 128;
  return windowWidth < windowHeight ? windowWidth : windowHeight;
};

class Main extends Component {
  componentDidMount = () => {
    window.addEventListener('resize', this.updateDimensions);
  };

  updateDimensions = () => {
    const { updateCanvasSize } = this.props;
    updateCanvasSize(getWindowSize());
  };

  render() {
    return (
      <main className="main">
        <section>
          <Tools />
          <ColorSelect />
        </section>
        <Frames />
        <Canvas />
        <section className="right-section">
          <Preview />
          <Layers />
        </section>
      </main>
    );
  }
}

Main.propTypes = {
  updateCanvasSize: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => {
  const { updateCanvasSize } = bindActionCreators(actions, dispatch);
  return {
    updateCanvasSize: (size) => {
      updateCanvasSize(size);
    },
  };
};

export default connect(
  undefined,
  mapDispatchToProps,
)(Main);
