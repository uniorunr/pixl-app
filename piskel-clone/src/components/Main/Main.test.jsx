import React from 'react';
import renderer from 'react-test-renderer';
import { JSDOM } from 'jsdom';
import Main from './Main';

const dom = new JSDOM();
global.document = dom.window.document;
global.window = dom.window;

function createNodeMock(element) {
  if (typeof element.type === 'string' && element.type === 'div') {
    const elm = document.createElement(element.type);
    elm.setAttribute('data-react-beautiful-dnd-drag-handle', '0');
    return elm;
  }
  if (element.type === 'input') {
    return {
      current: {
        value: 23,
      },
    };
  }
  return null;
}

it('renders correctly', () => {
  const options = { createNodeMock };
  const canvas = document.createElement('canvas');
  canvas.setAttribute('id', 'canvas');
  document.body.appendChild(canvas);

  const tree = renderer
    .create(
      <Main
        currToolId="pen"
        framesShortcuts={{}}
        layersShortcuts={{}}
        toolsData={{}}
        updateCurrentTool={() => {}}
      />,
      options,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
