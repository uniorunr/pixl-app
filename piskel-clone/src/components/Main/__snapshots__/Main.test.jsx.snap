// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="main"
>
  <section>
    <section
      className="tools-section"
    >
      <h2
        className="visually-hidden"
      >
        Tools Section
      </h2>
      <div
        className="tools"
      >
        <button
          className="tools__button tools_state_active"
          data-listener="true"
          id="pen"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <i
            className="fas fa-pen"
          />
        </button>
        <button
          className="tools__button"
          data-listener="true"
          id="paint-bucket"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <i
            className="fas fa-fill-drip"
          />
        </button>
        <button
          className="tools__button"
          data-listener="true"
          id="paint-same-pixels"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <i
            className="fas fa-paint-roller"
          />
        </button>
        <button
          className="tools__button"
          data-listener="true"
          id="choose-color"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <i
            className="fas fa-eye-dropper"
          />
        </button>
        <button
          className="tools__button"
          data-listener="true"
          id="eraser"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <i
            className="fas fa-eraser"
          />
        </button>
      </div>
    </section>
    <section
      className="color-select-section"
    >
      <div
        className="colors"
      >
        <div>
          <input
            className="colors__input-primary"
            defaultValue="#000000"
            onInput={[Function]}
            type="color"
          />
          <div
            className="colors__primary"
            style={
              Object {
                "backgroundColor": "#000000",
              }
            }
          />
        </div>
        <div>
          <input
            className="colors__input-secondary"
            defaultValue="#ffffff"
            onInput={[Function]}
            type="color"
          />
          <div
            className="colors__secondary"
            style={
              Object {
                "backgroundColor": "#ffffff",
              }
            }
          />
        </div>
      </div>
    </section>
  </section>
  <section
    className="frames-section"
  >
    <div
      className="frame frame0"
      data-active={true}
      data-id={0}
      role="presentation"
    >
      <canvas
        className="frame__canvas frame__canvas_active"
        height="128"
        id="frame0"
        onClick={[Function]}
        width="128"
      />
      <span
        className="frame__index"
      >
        1
      </span>
      <button
        className="frame__duplicate"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fas fa-copy"
        />
      </button>
    </div>
    <button
      className="frames-section__add-frame-button"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-plus"
      />
      <span>
        Add
      </span>
    </button>
  </section>
  <section
    className="canvas-section"
  >
    <canvas
      className="canvas-section__canvas"
      height={512}
      id="canvas"
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      width={512}
    />
  </section>
  <section
    className="preview-section"
  >
    <div
      className="preview"
    >
      <canvas
        className="preview__canvas"
        height="150"
        id="preview-canvas"
        width="150"
      />
      <button
        className="preview__fullscreen-button"
        type="button"
      >
        <i
          className="fas fa-expand"
        />
      </button>
      <div
        className="fps-control"
      >
        <label
          htmlFor="fps-control__input"
        >
          <span />
          <span>
            FPS
          </span>
          <input
            className="fps-control__input"
            id="fps-control__input"
            max="24"
            min="1"
            name="volume"
            onInput={[Function]}
            step="1"
            type="range"
          />
        </label>
      </div>
      <form
        className="canvas-size-controls"
        name="canvas-size-controls"
      >
        <label
          htmlFor="canvas-32"
        >
          32
          <input
            defaultChecked={true}
            id="canvas-32"
            name="radio"
            onChange={[Function]}
            type="radio"
            value="32"
          />
        </label>
        <label
          htmlFor="canvas-64"
        >
          64
          <input
            id="canvas-64"
            name="radio"
            onChange={[Function]}
            type="radio"
            value="64"
          />
        </label>
        <label
          htmlFor="canvas-128"
        >
          128
          <input
            id="canvas-128"
            name="radio"
            onChange={[Function]}
            type="radio"
            value="128"
          />
        </label>
      </form>
    </div>
  </section>
</main>
`;
